### DISCUSSION TITLE
Failing to install the my project in the virtual environment

### DISCUSSION THREAD

**Discussion by Alwoch** [pallets/flask#5151] [Q&A] [answered]
Hello everyone, I'm following the flask tutorial in the flask docs and stuck on [trying to make the project installable](https://flask.palletsprojects.com/en/2.3.x/tutorial/install/). I have added the pyproject.toml and MANIFEST.in as indicated in the tutorial but when I run `pip install -e`  to install the project in my virtual environment, I get this error:

`Usage:
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

-e option requires 1 argument`

I'm not quite sure what positional argument to add but when I add the application package (blog) as a positional argument, I get this error:

`Obtaining file:///D:/flask/tut/blog
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... error
  error: subprocess-exited-with-error

  × Getting requirements to build editable did not run successfully.
  │ exit code: 1
  ╰─> [14 lines of output]
      error: Multiple top-level packages discovered in a flat-layout: ['static', 'templates'].
     
      To avoid accidental inclusion of unwanted files or directories,
      setuptools will not proceed with this build.
     
      If you are trying to create a single distribution with multiple packages
      on purpose, you should not rely on automatic discovery.
      Instead, consider the following options:
     
      1. set up custom discovery (`find` directive with `include` or `exclude`)
      2. use a `src-layout`
      3. explicitly set `py_modules` or `packages` with a list of names
     
      To find more information, look for "package discovery" on setuptools docs.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: subprocess-exited-with-error

× Getting requirements to build editable did not run successfully.
│ exit code: 1
╰─> See above for output.

note: This error originates from a subprocess, and is likely not a problem with pip`

That was after I temporarily moved the pyproject.toml and MANIFEST.in into the package

Also I'm using windows and you can look at my project [here](https://github.com/Alwoch/flask_blog)

**Comment by umair313** [comment]
try `pip install -e .` make sure you are in the root director.

**Reply by Alwoch** [reply]
Hello @umair313 , I ran that from the root directory and It still failed

**Reply by umair313** [reply]
have you installed the requirements?

**Reply by umair313** [reply]
this works with python 3..11.3

python -m pip install -U pip setuptools wheel
pip install -r requirements/dev.txt
pip install -e .

**Comment by ThiefMaster** [comment]
Using Windows for web dev is already a poor choice. I seriously recommend you to consider using at least WSL2 instead of doing this stuff directly on Windows. vscode remote still lets you run your editor on windows, but the stuff would not run on windows itself that way.

Anyway, are you sure you did `pip install -e .` and not just `pip install -e`?

**Comment by EliseRemise** [comment, marked as answer]
I ran into this problem following the flaskr tutorial. The problem lies in the other directories in the root folder created earlier in the tutorial (instances and tests).  This confuses  pip install as well as build automation.  You are offered the solutions 1-3 but those are not very clear on howto if you are new to creating packages and wheels in python.  Easiest is number 3 by adding the following lines to your pyproject.toml file:

```
[tool.setuptools]
packages = ["flaskr"]
```

Now pip install and build have the package explicitly defined an do there thing without overthinking the presence of other directories. Maybe someone can add this to the tutorial so newcomers don't get confused ?

**Reply by Alwoch** [reply]
Thank you for this @EliseRemise I totally forgot to add this update too and its what worked for me. Thank you
